### PR TITLE
NSS TLS

### PR DESCRIPTION
**Pull request by ahf**
This PR tracks review for https://trac.torproject.org/projects/tor/ticket/26819

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -64,6 +66,16 @@ AM_CONDITIONAL(DISABLE_ASSERTS_IN_UNIT_TESTS, test "x$enable_asserts_in_tests" =
 AM_CONDITIONAL(LIBFUZZER_ENABLED, test "x$enable_libfuzzer" = "xyes")
 AM_CONDITIONAL(OSS_FUZZ_ENABLED, test "x$enable_oss_fuzz" = "xyes")
 AM_CONDITIONAL(USE_RUST, test "x$enable_rust" = "xyes")
+AM_CONDITIONAL(USE_NSS, test "x$enable_nss" = "xyes")
+AM_CONDITIONAL(USE_OPENSSL, test "x$enable_nss" != "xyes")
+
+if test "x$enable_nss" = "xyes"; then
+  AC_DEFINE(ENABLE_NSS, 1,
+  	    [Defined if we're building with NSS in addition to OpenSSL.])
```
Comment (ahf) [inline configure.ac]: The wording here is misleading: you can build Tor with NSS only.

File: src/core/mainloop/connection.c [hunk 2]
```diff
@@ -449,6 +449,20 @@ connection_new(int type, int socket_family)
   }
 }
 
+static void
+connection_close_and_invalidate_socket(connection_t *conn)
+{
+  if (connection_speaks_cells(conn)) {
+    or_connection_t *or_conn = TO_OR_CONN(conn);
+    tor_tls_free(or_conn->tls);
+    or_conn->tls = NULL;
```
Comment (ahf) [inline src/core/mainloop/connection.c]: `tor_tls_free()` should set `or_conn->tls` to `NULL`.

File: src/lib/crypt_ops/crypto_init.c [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+#include "lib/crypt_ops/crypto_curve25519.h"
+#include "lib/crypt_ops/crypto_dh.h"
+#include "lib/crypt_ops/crypto_ed25519.h"
+#include "lib/crypt_ops/crypto_openssl_mgt.h"
+#include "lib/crypt_ops/crypto_nss_mgt.h"
+#include "lib/crypt_ops/crypto_rand.h"
+
+#include "siphash.h"
+
+/** Boolean: has OpenSSL's crypto been initialized? */
```
Comment (ahf) [inline src/lib/crypt_ops/crypto_init.c]: We should probably remove the mentioning of OpenSSL here.

File: src/lib/crypt_ops/crypto_init.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+static int crypto_early_initialized_ = 0;
+
+/** Boolean: has OpenSSL's crypto been initialized? */
```
Comment (ahf) [inline src/lib/crypt_ops/crypto_init.c]: Same here. Remove OpenSSL from this comment.

File: src/lib/tls/tortls.c [hunk 5]
```diff
@@ -1856,407 +379,31 @@ tor_tls_free_(tor_tls_t *tls)
     size_t r,w;
     tor_tls_get_n_raw_bytes(tls,&r,&w); /* ensure written_by_tls is updated */
   }
-#ifdef SSL_set_tlsext_host_name
-  SSL_set_tlsext_host_name(tls->ssl, NULL);
-#endif
-  SSL_free(tls->ssl);
+  tor_tls_impl_free_(tls->ssl);
```
Comment (ahf) [inline src/lib/tls/tortls.c]: Why not have a `FREE_AND_NULL()` variant of this?

File: src/lib/tls/tortls_nss.c [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+ * \brief Wrapper functions to present a consistent interface to
+ * TLS and SSL X.509 functions from NSS.
+ **/
+
+#include "orconfig.h"
+
+#define TORTLS_PRIVATE
+#define TOR_X509_PRIVATE
+
+#ifdef _WIN32 /*wrkard for dtls1.h >= 0.9.8m of "#include <winsock.h>"*/
```
Comment (ahf) [inline src/lib/tls/tortls_nss.c]: I know this comes from a code movement, but this comment makes no sense to me at all. Maybe we should fix that afterwards.

File: src/core/mainloop/connection.c (hunk 2 above)
Comment (nmathewson) [reply src/core/mainloop/connection.c]: Right; the compiler should remove this redundancy for us.

File: configure.ac (hunk 1 above)
Comment (nmathewson) [reply configure.ac]: Fixed in 2f0eaaf89c5883bb63ed5734388bff96503be986

File: src/lib/crypt_ops/crypto_init.c (hunk 3 above)
Comment (nmathewson) [reply src/lib/crypt_ops/crypto_init.c]: done in 3730b9b85d52f9

File: src/lib/crypt_ops/crypto_init.c (hunk 4 above)
Comment (nmathewson) [reply src/lib/crypt_ops/crypto_init.c]: done in 3730b9b85d52f9

File: src/lib/tls/tortls.c (hunk 5 above)
Comment (nmathewson) [reply src/lib/tls/tortls.c]: tls_impl_free_ and tls_x509_impl_free_ are tiny wrappers around the underlying types.

File: src/lib/tls/tortls_nss.c (hunk 6 above)
Comment (nmathewson) [reply src/lib/tls/tortls_nss.c]: aa554cfaf48e03 revises this for the openssl case, and removes the comment for the nss case.

**Comment by ahf** [comment]
This has already been merged. Closing.

**Pull request closed without merging**